README.md
# PHP accessories

[![PHP Version](https://img.shields.io/packagist/php-v/alecrabbit/php-accessories.svg)](https://php.net)
[![Build Status](https://travis-ci.org/alecrabbit/php-accessories.svg?branch=master)](https://travis-ci.org/alecrabbit/php-accessories)
[![Appveyor Status](https://img.shields.io/appveyor/ci/alecrabbit/php-accessories.svg?label=appveyor)](https://ci.appveyor.com/project/alecrabbit/php-accessories/branch/master)
[![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/alecrabbit/php-accessories/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/alecrabbit/php-accessories/?branch=master)
[![Code Coverage](https://scrutinizer-ci.com/g/alecrabbit/php-accessories/badges/coverage.png?b=master)](https://scrutinizer-ci.com/g/alecrabbit/php-accessories/?branch=master)
[![Total Downloads](https://poser.pugx.org/alecrabbit/php-accessories/downloads)](https://packagist.org/packages/alecrabbit/php-accessories)

[![Latest Stable Version](https://poser.pugx.org/alecrabbit/php-accessories/v/stable)](https://packagist.org/packages/alecrabbit/php-accessories)
[![Latest Stable Version](https://img.shields.io/packagist/v/alecrabbit/php-accessories.svg)](https://packagist.org/packages/alecrabbit/php-accessories)
[![Latest Unstable Version](https://poser.pugx.org/alecrabbit/php-accessories/v/unstable)](https://packagist.org/packages/alecrabbit/php-accessories)

[![License](https://poser.pugx.org/alecrabbit/php-accessories/license)](https://packagist.org/packages/alecrabbit/php-accessories)
[![Average time to resolve an issue](http://isitmaintained.com/badge/resolution/alecrabbit/php-accessories.svg)](http://isitmaintained.com/project/alecrabbit/php-accessories "Average time to resolve an issue")
[![Percentage of issues still open](http://isitmaintained.com/badge/open/alecrabbit/php-accessories.svg)](http://isitmaintained.com/project/alecrabbit/php-accessories "Percentage of issues still open")

### Installation
```bash
composer require alecrabbit/php-accessories
```

### Usage
for details see [examples](https://github.com/alecrabbit/php-accessories/tree/master/examples)


### Features

##### Caller::class 
Gets a caller `Class::method()` or `function()` or `Undefined`
```php
$caller = Caller::get() // object(AlecRabbit\Accessories\Caller\CallerData)
```
> Note: `CallerData::class` can be casted to string
```php
if($wrongArguments) {
 throw new \RuntimeException(Caller::get() . ' provided wrong arguments'); 
}
```
You can set your custom formatter for string casting:
```php
$formatter = new CustomFormatter($options);
Caller::setFormatter($formatter);
```
> Note: `CustomFormatter::class` should implement `CallerDataFormatterInterface`

##### Circular::class
Helper class to get values in a circle
```php
$c = new Circular([1, 2, 3]);
$value = $c(); // int(1) invoke 
$value = $c->value(); // int(2) get value by method
... 
$c(); // int(3)
...
$c(); // int(1)
```
> Note: `Circular::__construct` can accept `array`, `Rewindable` or callable which returns `\Generator` 

##### Rewindable::class
Rewindable generator helper class
```php
$r = new Rewindable($genFunc); // $genFunc is a callable and returns \Generator
iterator_to_array($r);
$r->rewind();
```

##### G::class 
Contains generator functions
```php
$r = G::range(1, 3);  // object(Generator)
```

##### R::class 
Contains rewindable generator functions
```php
$r = R::range(1, 3); // object(AlecRabbit\Accessories\Rewindable)
iterator_to_array($r);
$r->rewind();
```

##### Pretty::class 
String formatter, e.g. percent, bytes and time(seconds, microseconds, nanoseconds)
```php
Pretty::bytes(10584760, 'mb'); // string(7) "10.09MB"
Pretty::time(0.214); // string(5) "214ms"
Pretty::percent(0.214);  // string(6) "21.40%"

Pretty::seconds(0.214); // string(5) "214ms"

Pretty::milliseconds(214); // string(5) "214ms"

Pretty::useconds(3212); // string(5) "3.2ms"
Pretty::useconds(12); // string(5) "12μs"
// alias for useconds
Pretty::microseconds(12); // string(5) "12μs"

Pretty::nanoseconds(10485); // string(7) "10.5μs"
Pretty::nanoseconds(105); // string(7) "105ns"
```
> Note: time formatting methods of `Pretty::class` named by units they are accepting

##### MemoryUsage::class
Helper class to get memory usage
```php
$report = MemoryUsage::reportStatic('mb');
echo $report . PHP_EOL;
// Memory: 0.75MB(32.73MB) Real: 2.00MB(34.00MB)
```
You can set your custom formatter for string casting:
```php
$formatter = new CustomFormatter($options);
MemoryUsage::setFormatter($formatter);
```
> Note: `CustomFormatter::class` should implement `MemoryUsageReportFormatterInterface`

> Note: Parameter `$options` has no effect on `MemoryUsageReportFormatter`


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

---

<a name="unreleased"></a>
## [Unreleased]

### Changed
- bumped dependency version

### Fixed
- tests' dataProviders data
- underlying package issue


<a name="0.12.1"></a>
## [0.12.1] - 2019-09-15
### Changed
- bumped deps version


<a name="0.12.0"></a>
## [0.12.0] - 2019-07-29
### Add
- Memory usage formatter diff string output
- test negative bytes values with `Pretty::bytes`


<a name="0.11.0"></a>
## [0.11.0] - 2019-07-29
### Changed
- `MemoryUsage::reportStatic`  renamed to `MemoryUsage::getReport`


<a name="0.10.0"></a>
## [0.10.0] - 2019-07-29
### Add
- `MemoryUsage::diff` function

### Changed
- Memory usage formatter string output

### Fix
- .gitattributes


<a name="0.9.1"></a>
## [0.9.1] - 2019-06-17
### Fix
- .gitattributes


<a name="0.9.0"></a>
## [0.9.0] - 2019-06-12

<a name="0.8.2-BETA.1"></a>
## [0.8.2-BETA.1] - 2019-04-29

<a name="0.8.1-BETA.1"></a>
## [0.8.1-BETA.1] - 2019-04-29

<a name="0.8.0-BETA.1"></a>
## [0.8.0-BETA.1] - 2019-04-28

<a name="0.7.1"></a>
## [0.7.1] - 2019-04-24

<a name="0.7.0"></a>
## [0.7.0] - 2019-04-24

<a name="0.6.7"></a>
## [0.6.7] - 2019-04-24

<a name="0.6.6"></a>
## [0.6.6] - 2019-04-24

<a name="0.6.5"></a>
## [0.6.5] - 2019-04-24

<a name="0.6.4"></a>
## [0.6.4] - 2019-04-12

<a name="0.6.3"></a>
## [0.6.3] - 2019-04-12

<a name="0.6.2"></a>
## [0.6.2] - 2019-04-11

<a name="0.6.1"></a>
## [0.6.1] - 2019-04-11

<a name="0.6.0"></a>
## [0.6.0] - 2019-04-11

<a name="0.5.3"></a>
## [0.5.3] - 2019-04-11

<a name="0.5.2"></a>
## [0.5.2] - 2019-04-04

<a name="0.5.1"></a>
## [0.5.1] - 2019-04-04

<a name="0.5.0"></a>
## [0.5.0] - 2019-03-03

<a name="0.4.3"></a>
## [0.4.3] - 2019-03-01

<a name="0.2.5"></a>
## [0.2.5] - 2019-02-10

<a name="0.1.1"></a>
## [0.1.1] - 2019-01-07

<a name="0.0.17"></a>
## 0.0.17 - 2018-12-14

[Unreleased]: https://github.com/alecrabbit/php-accessories/compare/0.12.1...HEAD
[0.12.1]: https://github.com/alecrabbit/php-accessories/compare/0.12.0...0.12.1
[0.12.0]: https://github.com/alecrabbit/php-accessories/compare/0.11.0...0.12.0
[0.11.0]: https://github.com/alecrabbit/php-accessories/compare/0.10.0...0.11.0
[0.10.0]: https://github.com/alecrabbit/php-accessories/compare/0.9.1...0.10.0
[0.9.1]: https://github.com/alecrabbit/php-accessories/compare/0.9.0...0.9.1
[0.9.0]: https://github.com/alecrabbit/php-accessories/compare/0.8.2-BETA.1...0.9.0
[0.8.2-BETA.1]: https://github.com/alecrabbit/php-accessories/compare/0.8.1-BETA.1...0.8.2-BETA.1
[0.8.1-BETA.1]: https://github.com/alecrabbit/php-accessories/compare/0.8.0-BETA.1...0.8.1-BETA.1
[0.8.0-BETA.1]: https://github.com/alecrabbit/php-accessories/compare/0.7.1...0.8.0-BETA.1
[0.7.1]: https://github.com/alecrabbit/php-accessories/compare/0.7.0...0.7.1
[0.7.0]: https://github.com/alecrabbit/php-accessories/compare/0.6.7...0.7.0
[0.6.7]: https://github.com/alecrabbit/php-accessories/compare/0.6.6...0.6.7
[0.6.6]: https://github.com/alecrabbit/php-accessories/compare/0.6.5...0.6.6
[0.6.5]: https://github.com/alecrabbit/php-accessories/compare/0.6.4...0.6.5
[0.6.4]: https://github.com/alecrabbit/php-accessories/compare/0.6.3...0.6.4
[0.6.3]: https://github.com/alecrabbit/php-accessories/compare/0.6.2...0.6.3
[0.6.2]: https://github.com/alecrabbit/php-accessories/compare/0.6.1...0.6.2
[0.6.1]: https://github.com/alecrabbit/php-accessories/compare/0.6.0...0.6.1
[0.6.0]: https://github.com/alecrabbit/php-accessories/compare/0.5.3...0.6.0
[0.5.3]: https://github.com/alecrabbit/php-accessories/compare/0.5.2...0.5.3
[0.5.2]: https://github.com/alecrabbit/php-accessories/compare/0.5.1...0.5.2
[0.5.1]: https://github.com/alecrabbit/php-accessories/compare/0.5.0...0.5.1
[0.5.0]: https://github.com/alecrabbit/php-accessories/compare/0.4.3...0.5.0
[0.4.3]: https://github.com/alecrabbit/php-accessories/compare/0.2.5...0.4.3
[0.2.5]: https://github.com/alecrabbit/php-accessories/compare/0.1.1...0.2.5
[0.1.1]: https://github.com/alecrabbit/php-accessories/compare/0.0.17...0.1.1
- the format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
- this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html)


examples/pretty.php
<?php

require_once __DIR__ . '/../vendor/autoload.php';

use const \AlecRabbit\Helpers\Constants\UNIT_MICROSECONDS;
use AlecRabbit\Pretty;

var_dump(Pretty::bytes(10485760, 'GB')); // string(6) "0.01GB"
var_dump(Pretty::bytes(10584760, 'MB')); // string(7) "10.09MB"
var_dump(Pretty::bytes(10485760, 'KB')); // string(10) "10240.00KB"
echo PHP_EOL;
var_dump(Pretty::time(0.001048576454530, UNIT_MICROSECONDS, 6)); // string(11) "1048.576μs"
var_dump(Pretty::time(0.0000000021)); // string(5) "2.1ns"
var_dump(Pretty::time(0.214)); // string(5) "214ms"
var_dump(Pretty::time(10485760)); // string(9) "2912.711h"
var_dump(Pretty::time(3212)); // string(6) "53.53m"
var_dump(Pretty::time(time())); // find out yourself :)
echo PHP_EOL;
var_dump(Pretty::nanoseconds(10485)); // string(7) "10.5μs"
var_dump(Pretty::seconds(0.214)); // string(5) "214ms"
var_dump(Pretty::useconds(3212)); // string(5) "3.2ms"
var_dump(Pretty::useconds(12)); // string(5) "12μs"
echo PHP_EOL;
var_dump(Pretty::percent(0.214)); // string(6) "21.40%"

